modules/core/include/opencv2/core/cuda.hpp

```diff
@@ -337,6 +337,195 @@ class CV_EXPORTS_W GpuMat
+    function may return even if the copy operation is not finished.
+
+    The copy operation may be overlapped with operations in other non-default streams if \p stream is
+    not the default stream and \p dst is HostMem allocated with HostMem::PAGE_LOCKED option.
+    */
+    void download(OutputArray dst, Stream& stream) const;
+
+    //! returns deep copy of the GpuMatND, i.e. the data is copied
+    GpuMatND clone() const;
+
+    //! copies the GpuMatND content to device memory (Blocking call)
+    void copyTo(OutputArray dst) const;
+
+    //! copies the GpuMatND content to device memory (Non-Blocking call)
+    void copyTo(OutputArray dst, Stream& stream) const;
+
+    //! copies those GpuMatND elements to "m" that are marked with non-zero mask elements (Blocking call)
+    void copyTo(OutputArray dst, InputArray mask) const;
+
+    //! copies those GpuMatND elements to "m" that are marked with non-zero mask elements (Non-Blocking call)
+    void copyTo(OutputArray dst, InputArray mask, Stream& stream) const;
+
+    //! sets some of the GpuMatND elements to s (Blocking call)
+    GpuMatND& setTo(Scalar s);
+
+    //! sets some of the GpuMatND elements to s (Non-Blocking call)
+    GpuMatND& setTo(Scalar s, Stream& stream);
+
+    //! sets some of the GpuMatND elements to s, according to the mask (Blocking call)
+    GpuMatND& setTo(Scalar s, InputArray mask);
+
+    //! sets some of the GpuMatND elements to s, according to the mask (Non-Blocking call)
+    GpuMatND& setTo(Scalar s, InputArray mask, Stream& stream);
+
+    //! converts GpuMatND to another datatype (Blocking call)
+    void convertTo(OutputArray dst, int rtype) const;
+
+    //! converts GpuMatND to another datatype (Non-Blocking call)
+    void convertTo(OutputArray dst, int rtype, Stream& stream) const;
+
+    //! converts GpuMatND to another datatype with scaling (Blocking call)
+    void convertTo(OutputArray dst, int rtype, double alpha, double beta = 0.0) const;
+
+    //! converts GpuMatND to another datatype with scaling (Non-Blocking call)
+    void convertTo(OutputArray dst, int rtype, double alpha, Stream& stream) const;
+
+    //! converts GpuMatND to another datatype with scaling (Non-Blocking call)
+    void convertTo(OutputArray dst, int rtype, double alpha, double beta, Stream& stream) const;
+
+    void assignTo(GpuMatND& m, int type=-1) const;
+
+    //! returns pointer to y-th row
+    uchar* ptr(int y = 0);
+    const uchar* ptr(int y = 0) const;
+
+    //! template version of the above method
+    template<typename _Tp> _Tp* ptr(int y = 0);
+    template<typename _Tp> const _Tp* ptr(int y = 0) const;
+
+    template <typename _Tp> operator PtrStepSz<_Tp>() const;
+    template <typename _Tp> operator PtrStep<_Tp>() const;
+
+    //! creates alternative GpuMatND header for the same data, with different
+    //! number of channels and/or different number of rows
+    GpuMatND reshape(int* new_shape) const;
```

Comment:
how will you figure out the number of dimensions?

Reply:
Ah! I forgot to replace this with `vector<int> new_shape`. Thanks for pointing out.